share/completions/meson.fish

```diff
@@ -1,52 +1,308 @@
-    ",sysconfdir,Sysconf data directory [etc]"
-
-for arg in $basic_arguments
-    set -l parts (string split , -- $arg)
-    if not string match -q "" -- $parts[1]
-        complete -c meson -s "$parts[1]" -l "$parts[2]" -d "$parts[3]"
+function __fish_meson_needs_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    argparse -s 'v/version' -- $cmd 2>/dev/null
+    or return 0
+    not set -q argv[1]
+end
+
+function __fish_meson_using_command
+    set -l cmd (commandline -opc)
+    set -e cmd[1]
+    test (count $cmd) -eq 0
+    and return 1
+    contains -- $cmd[1] $argv
+    and return 0
+end
+
+function __fish_meson_builddir
+    # Consider the value of -C option to detect the build directory
+    set -l cmd (commandline -opc)
+    argparse -i 'C=' -- $cmd
+    if set -q _flag_C
+        echo $_flag_C
     else
-        complete -c meson -l "$parts[2]" -d "$parts[3]"
+        echo .
     end
 end
 
-for arg in $dir_arguments
-    set -l parts (string split , -- $arg)
-    complete -c meson -l "$parts[2]" -d "$parts[3]" -xa '(__fish_complete_directories)'
+function __fish_meson_targets
+    set -l python (__fish_anypython); or return
+    meson introspect --targets (__fish_meson_builddir) | $python -S -c 'import json, sys
+data = json.load(sys.stdin)
+targets = set()
+for target in data:
+    targets.add(target["name"])
+for name in targets:
+    print(name)' 2>/dev/null
+end
+
+function __fish_meson_subprojects
+    set -l python (__fish_anypython); or return
+    meson introspect --projectinfo (__fish_meson_builddir) | $python -S -c 'import json, sys
+data = json.load(sys.stdin)
+for subproject in data["subprojects"]:
+    print(subproject["name"])' 2>/dev/null
+end
+
+function __fish_meson_tests
+    # --list option shows suites in a short form, e.g. if a test "gvariant"
+    # is present both in "glib:glib" and "glib:slow" suites, it will be shown
+    # in a list as "glib:glib+slow / gvariant". So, just filter out the first
+    # part and list all of the test names.
+    meson test -C (__fish_meson_builddir) --no-rebuild --list | string split -r -f1 ' / '
+end
+
+function __fish_meson_test_suites
+    set -l python (__fish_anypython); or return
+    meson introspect --tests (__fish_meson_builddir) | $python -S -c 'import json, sys
+data = json.load(sys.stdin)
+suites = set()
+for test in data:
+    suites.update(test["suite"])
+for name in suites:
+    print(name)' 2>/dev/null
+end
+
+function __fish_meson_help_commands
+    meson help --help | string match -g -r '^  {(.*)}' | string split ,
 end
 
-complete -c meson -s D -d "Set value of an option (-D foo=bar)"
+# Each meson command and subcommand has -h/--help option
+complete -c meson -s h -l help -d 'Show help'
+
+# global options
+complete -c meson -n __fish_meson_needs_command -s v -l version -d 'Show version number and exit'
+# implicit "meson setup" usage via just "meson" was deprecated a long time ago,
+# so no "meson setup" option completions in global scope
```

Comment:
I'm not an expert on what qualifies as "a long time ago" and what doesn't, but FWIW this was news to me. It's going to take me a long time to stop doing `meson build` or `meson --buildtype debug debug`

Some git sleuthing shows the deprecation was only added to the codebase August 2022 and shipped in meson 0.64.0, which didn't ship until November of 2022. Debian stable is shipping with meson 0.56.2, Ubuntu's bleeding edge is shipping meson 0.63 while 22.04 LTS is shipping 0.61, and 20.04 is shipping 0.53.

If these were new completions altogether, I'd say it's fine. But the old completions supported top-level setup (which, deprecated or not, was how meson themselves were using it in the docs for the longest time), so this is actually a regression in our completions.

I would just change `-n '__fish_meson_using_command setup'` to `-n '__fish_meson_needs_command || __fish_meson_using_command setup'` so that you can automatically support both?

Reply:
There's a lot more context in the relevant Meson commit: https://github.com/mesonbuild/meson/commit/3c7ab542c0c4770241eae149b0d4cd8de329aee0

The commit message calls out the fact that since Meson 0.42.0 (which is indeed "a long time ago") `meson builddir/` is silently rewritten to `meson setup builddir/` -- and also observes that the fact that it wasn't deprecated way back in 0.42.0 was an *accident*, because we deprecated the other entry points via wrapper scripts but forgot to do anything to the main executable.

Reply:
Sorry, I meant to link that commit in my response.

> [...] and also observes that the fact that it wasn't deprecated way back in 0.42.0 was an accident

"If a tree falls in the forest and no one hears it...."

> So, in fact, bringing setup option completions into a global scope seems to be harmful, as more people might discover via completions that it works and start to use it in this way.

This is a problem for meson, but fish serves to facilitate a *user's* workflow. The commands and options are supported today and people use them today. The interests of the developers of the tools we provide completions for and the interests of our users are almost always in lock-step but sometimes they do not align perfectly.

If you do not wish to add the completions, that's fine. Fish might still support them anyway.

Reply:
I do not care one way or the other (neither for Meson nor on a personal level), I'm just pointing out some context in case you're worried about personally relying on `meson setup`.

> It's going to take me a long time to stop doing `meson build` or `meson --buildtype debug debug`

etc.

Reply:
I understand and appreciate your point. Thanks for stopping by and providing a definitive answer there and thanks for the work you do on meson!

Reply:
> This is a problem for meson, but fish serves to facilitate a user's workflow. The commands and options are supported today and people use them today. The interests of the developers of the tools we provide completions for and the interests of our users are almost always in lock-step but sometimes they do not align perfectly.

Yeah, I think that I got your point. I just thought about these completions more from the meson side, not the fish side. If they were in the meson repo, then it would be perfectly fine not to include these global options. But completions are in the fish repo, and fish cannot assume much about which version of meson the user has. So, you are right, and it's better to support them here.

Reply:
@mqudsi

No worries, that's the way I interpreted it too. :)

Reply:
Thanks, @dseight. I'm all for being able to deprecate or change interfaces, but we just want to be nice to everyone :)

This looks good. I'll take a final look and merge if I don't notice anything else!